tools/gen-random-bout/gen-random-bout.cpp

```diff
@@ -22,116 +25,276 @@
 #define stat64 stat
 #endif
 
-// --sym-args 0 1 10 --sym-args 0 2 2 --sym-files 1 8 --sym-stdout
-static int getint(char *i) {
-  if(!i) {
-    fprintf(stderr, "ran out of arguments!\n");
-    assert(i);
+#define SMALL_BUFFER_SIZE 64 // To hold "arg<N>" string, temporary
+                             // filename, etc.
+
+#define MAX_FILE_SIZES 256
+
+static void error_exit(const char *fmt, ...) {
+  va_list args;
+
+  va_start(args, fmt);
+  vfprintf(stderr, fmt, args);
+  va_end(args);
+
+  exit(1);
+}
+
+static unsigned get_unsigned(char *i) {
+  if (!i) {
+    error_exit("ran out of arguments!\n");
   }
-  return atoi(i);
+  long int n = strtol(i, NULL, 10);
+  if (n < 0 || n == LONG_MIN || n == LONG_MAX) {
+    error_exit("%s:%d: Error in conversion to unsigned: %s\n", __FILE__,
+               __LINE__, i);
+  }
+  return (unsigned)n;
 }
 
 #define MAX 64
-static void push_obj(KTest *b, const char *name, unsigned non_zero_bytes, 
-                     unsigned total_bytes) {
+static void push_random_obj(KTest *b, const char *name, unsigned non_zero_bytes,
+                            unsigned total_bytes) {
   KTestObject *o = &b->objects[b->numObjects++];
   assert(b->numObjects < MAX);
 
-  o->name = strdup(name);
+  if ((o->name = strdup(name)) == NULL) {
+    error_exit("%s:%d: strdup() failure\n", __FILE__, __LINE__);
+  }
   o->numBytes = total_bytes;
-  o->bytes = (unsigned char *)malloc(o->numBytes);
+  if ((o->bytes = (unsigned char *)malloc(o->numBytes)) == NULL) {
+    error_exit("%s:%d: malloc() failure\n", __FILE__, __LINE__);
+  }
 
   unsigned i;
-  for(i = 0; i < non_zero_bytes; i++)
-    o->bytes[i] = random();
+  for (i = 0; i < non_zero_bytes; i++) {
+    o->bytes[i] = random() % 255 + 1;
+  }
 
-  for(i = non_zero_bytes; i < total_bytes; i++)
+  for (i = non_zero_bytes; i < total_bytes; i++)
     o->bytes[i] = 0;
 }
 
+static void push_obj(KTest *b, const char *name, unsigned total_bytes,
+                     unsigned char *content) {
+  KTestObject *o = &b->objects[b->numObjects++];
+  assert(b->numObjects < MAX);
+
+  if ((o->name = strdup(name)) == NULL) {
+    error_exit("%s:%d: strdup() failure\n", __FILE__, __LINE__);
+  }
+  o->numBytes = total_bytes;
+  if ((o->bytes = (unsigned char *)malloc(total_bytes)) == NULL) {
+    error_exit("%s:%d: malloc() failure\n", __FILE__, __LINE__);
+  }
+  memcpy(o->bytes, content, total_bytes);
+}
 
 static void push_range(KTest *b, const char *name, unsigned value) {
```

Comment:
With `push_obj` above `push_range` is redundant, isn't it? Just pass sizeof(value), &value...

Reply:
@251 That is again part of the old code you are complaining about, but no problem I'll fix it.

Reply:
@251 Instead of deleting `push_range`, I made it call `push_obj`. This is so that `push_range` can be called with a constant unsigned integer as an argument (see the call to `push_range` in `main`), and the allocation for that value to be passed on to `push_obj` will be automatically made to store the argument when `push_range` is called.